File: app/src/main/java/com/kml/adapters/WorkAdapter.java
package com.kml.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.kml.R;
import com.kml.data.models.Work;
import com.kml.holders.WorkHolder;


import java.util.ArrayList;
import java.util.List;

public class WorkAdapter extends RecyclerView.Adapter<WorkHolder>
{
    List<Work> works = new ArrayList<>();
    private WorkAdapter.OnItemClickListener listener;
    ProgressBar progressBar;

    @NonNull
    @Override
    public WorkHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
    {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.single_item_work_history, parent, false);
        return new WorkHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull WorkHolder holder, int position)
    {
        Work work = works.get(position);
        holder.workHistoryName.setText(work.getWorkName());
        holder.workHistoryDescription.setText(work.getWorkDescription());
        holder.workHistoryDate.setText(work.getWorkDate());
        holder.itemView.setOnClickListener(v -> {
            if(listener!=null && position != RecyclerView.NO_POSITION)
            {
                listener.OnItemClick(works.get(position));
            }
        });
    }

    @Override
    public int getItemCount()
    {
        return works.size();
    }

    public void setWorks(List<Work> works)
    {
        this.works = works;
        notifyDataSetChanged();
    }

    public void setProgressBar(ProgressBar progressBar)
    {
        this.progressBar = progressBar;
    }

    public interface OnItemClickListener
    {
        void OnItemClick(Work work);
    }

    public void setOnItemClickListener(WorkAdapter.OnItemClickListener listener)
    {
        this.listener = listener;
    }
}

File: app/src/main/java/com/kml/adapters/WorkAdapter.kt
package com.kml.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.kml.R
import com.kml.holders.EmptyWorkHolder
import com.kml.holders.WorkHolder
import com.kml.models.Work
import java.util.*

class WorkAdapter(private val onClickListener: (Work) -> Unit)
    : RecyclerView.Adapter<WorkHolder>() {
    private var worksConstant: List<Work> = listOf()
    private var works: List<Work> = listOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WorkHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.list_item_work_history, parent, false)
        return if (viewType == WORK_VIEW_TYPE) WorkHolder(itemView)
        else EmptyWorkHolder(itemView)
    }

    override fun onBindViewHolder(holder: WorkHolder, position: Int) {
        val work = works[position]
        holder.bind(work, onClickListener)
    }

    override fun getItemCount(): Int {
        return works.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (works[position].isEmpty()) EMPTY_WORK_VIEW_TYPE
        else WORK_VIEW_TYPE
    }

    fun updateWorks(newWorks: List<Work>) {
        works = newWorks + EMPTY_WORK
        if (worksConstant.isEmpty())
            worksConstant = newWorks
        notifyDataSetChanged()
    }

    fun filterWorksBy(title: String): List<Work> {
        val filteredWorks = worksConstant.filter {
            it.workName.toLowerCase(Locale.ROOT).contains(title.toLowerCase(Locale.ROOT))
        }
        works = filteredWorks + EMPTY_WORK
        notifyDataSetChanged()
        return filteredWorks
    }

    companion object {
        val EMPTY_WORK = Work("","","","","","","")

        const val EMPTY_WORK_VIEW_TYPE = 1
        const val WORK_VIEW_TYPE = 2
    }
}